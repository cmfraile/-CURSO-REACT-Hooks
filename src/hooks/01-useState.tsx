import { useState } from "react"

export const useStateHook = () => {

    const [{c1,c2,c3},setCounter] = useState<any>({c1:10,c2:20,c3:30});

    return(
        <>
        <h2>useState</h2>
        <code>{JSON.stringify({c1,c2,c3})}</code>
        <button className="btn btn-primary" onClick={() => { setCounter({c1:c1+1,c2,c3}) }}>+1</button>
        </>
    )
}