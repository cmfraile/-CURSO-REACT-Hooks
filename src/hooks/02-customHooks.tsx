import { useState } from "react"

const useCounter = (value:number = 10) => {
    const [ counter , setCounter ] = useState(value);
    return{counter}
}

export const useCustomHook = () => {

    const { counter } = useCounter();

    return(
        <>
        <h2>Custom hook</h2>

        <code>{counter}</code>

        <button className="btn btn-primary" >+1</button>
        <button className="btn btn-primary" >-1</button>
        <button className="btn btn-primary" >RESET</button>

        </>
    )
}