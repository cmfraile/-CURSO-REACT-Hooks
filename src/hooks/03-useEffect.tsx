import { useState , useEffect } from "react"

const Message = () => {

    useEffect(() => {
        console.log("montado");
        return () => {
            console.log("DESmontado");
        }
    })

    return(
        <>
            <h3>Usuario ya existe</h3>
        </>
    )
}

export const useEffectHook = () => {

    const [formState,setFormState] = useState({
        username:'',
        email:''
    });
    const { username , email } = formState;
    
    const onInputChange = ({target}:any) => {
        const { name , value } = target;
        setFormState({
            ...formState,
            [ name ]:value
        })
    }

    /*
    useEffect(() => {console.log('username!')},[username]);
    useEffect(() => {console.log('email!')},[email]);
    */


    

    return (
        <>
            <h2>UseEffectHook</h2>
            <input
                    type="text"
                    className="form-control iform"
                    placeholder="username"
                    name="username"
                    value={username}
                    onChange={onInputChange}
            />
            <input
                    type="text"
                    className="form-control iform"
                    placeholder="email"
                    name="email"
                    value={email}
                    onChange={onInputChange}
            />

            {(username === 'Tortilla') && <Message/>}

        </>
    )
}