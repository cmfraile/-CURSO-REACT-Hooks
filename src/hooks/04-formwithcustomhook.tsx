import { useState , useEffect } from "react"

const Message = () => {

    const [ xy , setXy ] = useState<{x:number,y:number}>({x:0,y:0});

    useEffect(() => {
        const onMouseMove = ({x,y}:{x:number,y:number}) => {setXy({x,y})};
        window.addEventListener('mousemove',onMouseMove);
        return () => {window.removeEventListener('mousemove',onMouseMove)};
    })

    return(<><h3>Usuario existe{JSON.stringify(xy)}</h3></>);
}

const useForma = (initialForm:any = {}) => {

    /*
        const [formState,setFormState] = useState({username:'',email:'',password:''});
        const formgenerator:any = {username:formState.username,email:formState.email,password:formState.password};
        const formgenerator:any = {username:formState.username,email:formState.email,password:formState.password};
    */

    const [formState,setFormState] = useState(initialForm);
    
    const onInputChange = ({target}:any) => {
        const { name , value } = target;
        setFormState({
            ...formState,
            [ name ]:value
        })
    }

    const onResetForm = () => { setFormState(initialForm) };
    
    return({formState,onInputChange,onResetForm});
}

export const Formwithcustomhook = () => {

    const {formState,onInputChange,onResetForm} = useForma({username:'',email:'',password:''})

    /*
    useEffect(() => {console.log('username!')},[username]);
    useEffect(() => {console.log('email!')},[email]);
    */

    return (
        <>
            <h2>Form with custom hook</h2>
            {Object.keys(formState).map((x:string) => {
                return(
                    <input
                        key={x}
                        type={(x === 'password') ? x : 'text'}
                        className="form-control iform"
                        placeholder={x}
                        name={x}
                        value={formState[x]}
                        onChange={onInputChange}
                    />
                )
            })}

            <button className="btn btn-primary" onClick={onResetForm}>RESET</button>

            {(formState.username === 'Tortilla') && <Message/>}

        </>
    )
}