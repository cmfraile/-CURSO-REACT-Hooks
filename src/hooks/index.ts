export * from './01-useState';
export * from './02-customHooks';
export * from './03-useEffect';
export * from './04-formwithcustomhook';
export * from './05-examples';
export * from './06-useRef';
export * from './07-useLayoutEffect';
export * from './08-memos';
export * from './09-memohook';
export * from './10-callbackhook';
export * from './tareamemo/Padre';