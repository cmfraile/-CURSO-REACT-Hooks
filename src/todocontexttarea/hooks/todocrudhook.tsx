import { useReducer ,useEffect, useState } from "react";
import { random } from "underscore";

export enum tiposacc {add = '[TODO] Add Todo',del = '[TODO] Del Todo',nuke = '[TODO] Nuke Todo',end = '[TODO] end TODO',edit = '[TODO] edit TODO'};
export interface todobj {id:number,todo:string,done:boolean};export interface action {type:tiposacc,payload:todobj} ;

export const todocraft = (todo:string):todobj => {return {id:(new Date().getTime() + random(0,10000)),todo,done:false}};
export const todoReducer = (state:todobj[]|[] = [],action:action) => {
    if(!action){return state}
    const { type , payload } = action ; const { add , nuke , del , end , edit } = tiposacc ;
    switch(type){
        case add : return [...state,payload];
        case del : return state.filter(x => x.id !== payload.id);
        case edit : return new Array().concat(state.filter(x => x.id !== payload.id),payload) 
        case end : return state.map(x => {if(x.id == payload.id){x.done = true ; return x}else{return x}});
        case nuke : return [];
        default : throw new Error();
    }
}

export const todocrudch = () => {
    const [ todos , todosDispatch ] = useReducer(todoReducer,[],() => {
        const caso = localStorage.getItem('todos');
        if(caso){return JSON.parse(caso)}else{return {logged:false}}
    });
    const [ edit , setEdit ] = useState<todobj|undefined>();
    useEffect(() => { localStorage.setItem('todos',JSON.stringify(todos))},[todos]);
    const todoCRUD = {
        addTODO:(todo:string) => {
            const action:action = {type:tiposacc.add,payload:todocraft(todo)};
            todosDispatch(action);
        },
        delTODO:(todo:todobj) => {
            const action:action = {type:tiposacc.del,payload:todo};
            todosDispatch(action);
        },
        endTODO:(todo:todobj) => {
            const action:action = {type:tiposacc.end,payload:todo};
            todosDispatch(action);
        },
        editTODO:(todo:todobj) => {
            const action:action = {type:tiposacc.edit,payload:todo};
            todosDispatch(action);
        },
        nukeTODO:() => {
            todosDispatch({type:tiposacc.nuke,payload:todocraft('nuke')})
        }
    }
    return({todos,todoCRUD,...todoCRUD,edit,setEdit});
}