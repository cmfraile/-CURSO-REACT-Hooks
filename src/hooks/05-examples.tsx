import { useEffect, useLayoutEffect, useRef, useState } from "react"
import { useCounter } from "./02-customHooks";
import { useLayoutEffectHook } from "./07-useLayoutEffect";

const useFetch = (url:string) => {
    const [ state , setState ] = useState({data:null,isLoading:true,error:null})
    const getFetch = async() => {
        setState({...state,isLoading:true});
        await(await fetch(url)).json()
        .then(data => {setState({data,isLoading:false,error:null})})
        .catch(error => {setState({data:null,isLoading:false,error})})
    }
    useEffect(() => { getFetch() },[url]);
    return({...state});
}

export const multipleCustomHooks = () => {

    const { data , isLoading , error } = useFetch('https://www.breakingbadapi.com/api/quotes');
    const { counter , sc } = useCounter();
    const { height , setHeight } = useLayoutEffectHook()

    const iref = useRef<HTMLParagraphElement>((<p/>).type);
    useLayoutEffect(() => {setHeight(iref.current.getBoundingClientRect().height)});
    //useEffect(() => {setHeight(iref.current.getBoundingClientRect().height)});
    
    
    return(
        <>
        <h1 style={{color:(error) ? 'red' : 'black'}} >{(error) ? 'ERROR!' : (isLoading) ? 'Cargando' : 'Breaking Bad Quotes'}</h1>
        <p ref={iref}>{(data) && data[counter]['quote']}</p>
        <p>Altura del quote : {height}</p>
        <button className="btn btn-primary" onClick={() => {sc('+')}}>+1</button>
        <button className="btn btn-primary" disabled={(counter == 0) ? true : false} onClick={() => {sc('-')}}>-1</button>
        <button className="btn btn-primary" disabled={(counter == 0) ? true : false} onClick={() => {sc('R')}}>RESET</button>
        </>
    )
}