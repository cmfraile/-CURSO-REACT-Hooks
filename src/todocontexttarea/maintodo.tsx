import { useContext, useEffect } from 'react';
import { useForma } from '../hooks/04-formwithcustomhook';
import { todoContext } from './context/todoContext';
import './todo.sass';

interface todobj {id:number,todo:string,done:boolean};

const TodoList = ({todos = []}:{todos:todobj[] | []}) => {
    console.log('render de lista');
    return(<><ul>{(todos) && todos
        .sort((a,b) => {
            const sorta = a.done ; const sortb = b.done;
            if(sorta == sortb){return 0}
            if(sorta){return 1} ; if(sortb){return -1}
            return 0;
        })
        .map(({id,todo,done}:todobj,i:number) => {
            const marginbool = ():boolean => {if(i == 0){return false} ; if(done && !todos[i-1].done){return true}else{return false}}
            return <TodoItem key={id} id={id} todo={todo} done={done} margin={marginbool()}/>})}</ul></>
        )};
const TodoItem = ({id,todo,done,margin}:any) => {
    const { setEdit } = useContext(todoContext);
    const { delTODO , endTODO } = useContext(todoContext);
    return(<><li    className={`litem ${(margin) ? 'margin' : ''}`} 
                    style={{textDecoration: (done) && 'line-through'}}
                    onDoubleClick={() => {setEdit({id,todo,done})}} 
                    key={id}>{todo}
    <button className='btn btn-danger' onClick={() => {delTODO({id,todo,done})}}></button>
    <button className='btn btn-primary' disabled={done} onClick={() => {endTODO({id,todo,done})}}></button>
    </li></>)
};
const TodoAdd = ({todos,onNewTodo,onEditTodo,onNuke}:{onNewTodo:any,onEditTodo:any,todos:todobj[],onNuke:any}) => {

    const {todo,onInputChange,onResetForm} = useForma({todo:''});
    const { edit , setEdit } = useContext<{edit:todobj,setEdit:any}>(todoContext);

    const validarmiddle = (todo:string):boolean => {
        const caso = todo.trim().toUpperCase();
        if(todo.trim().length <= 2){return false};
        if(todos.map(x => x.todo.toUpperCase()).includes(caso)){return false};
        return true
    }

    useEffect(() => {
        if(!edit){return}
        onInputChange({target:{name:'todo',value:edit.todo}});
    },[edit])
    
    return(
        <>
            <form
            onSubmit={(e) => {
                e.preventDefault();
                if(!validarmiddle(todo)){return}
                if(!edit){onNewTodo(todo.trim())}
                if(edit){onEditTodo({id:edit.id,todo,done:edit.done}) ; setEdit(undefined)}
                onResetForm();
            }}>
            <input
            className="form-control"
            type="text"
            name="todo"
            value={todo}
            placeholder={'aÃ±adir todo'}
            onChange={onInputChange}
            />
            </form>
            <button className="btn btn-danger" disabled={(todos.length == 0) && true} onClick={() => {if(confirm('esta seguro de mandar todo a la megamierda?')){onNuke() ; setEdit(undefined)}}}>NUKE</button>
            {(edit) && <code>{JSON.stringify(edit.todo)}</code>}
        </>
    )

}

export const TodoApp = () => {

    const { todos , addTODO , nukeTODO , editTODO } = useContext(todoContext);

    return (
        <>
        <h2>Todo app : {todos.length} tareas pendientes</h2>
        <div className="container">
            <div className="row">
                <div className="col"><TodoList todos={todos} /></div>
                <div className="col"><TodoAdd   onNewTodo={addTODO}
                                                onEditTodo={editTODO}
                                                onNuke={nukeTODO}
                                                todos={todos}/></div>
            </div>
        </div>
        </>
    )
}