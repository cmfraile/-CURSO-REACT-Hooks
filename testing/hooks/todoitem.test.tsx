import React from 'react';
import { TodoItem } from '../../src/todocontexttarea/maintodo';
import { render , renderHook, screen } from '@testing-library/react';

import { todocraft } from '../../src/todocontexttarea/hooks/todocrudhook';

const main = () => {

    test('snapshot de componente',() => {
        const { id , done , todo ,margin } = {...todocraft('tododeprueba'),margin:false};
        expect(render(<TodoItem id={id} todo={todo} done={done} margin={margin} />).container).toMatchSnapshot();
        screen.debug();
    })
    
    test('vamos a descomponer el todoitemhook',() => {
        const { id , done , todo ,margin } = {...todocraft('tododeprueba'),margin:false};
        const rHOOK = renderHook(() => <TodoItem id={id} todo={todo} done={done} margin={margin} />);
    });
    
    test('debe de tacharse un todo completado', () => {
        const { id , todo ,margin } = {...todocraft('tododeprueba'),margin:false};
        render(<TodoItem id={id} todo={todo} done={true} margin={margin} />);
        const litem = screen.getByRole('listitem');
        expect(litem.style['_values']['text-decoration']).toBe("line-through");
    });

}

describe('pruebas en el todoitem',main);