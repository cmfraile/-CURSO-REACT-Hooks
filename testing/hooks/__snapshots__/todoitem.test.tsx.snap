// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el todoitem snapshot de componente 1`] = `
<div>
  <li
    class="litem "
  >
    tododeprueba
    <button
      class="btn btn-danger"
    />
    <button
      class="btn btn-primary"
    />
  </li>
</div>
`;
